### ISSUE TITLE
"Single-shard transactions" do not respect the transaction's isolation level

### ISSUE THREAD

**Issue by lithp** [citusdata/citus#708]
As of [this commit](https://github.com/citusdata/citus/commit/5d525fba24e09085457e55f335cd13a09299e5fe) we can run queries on distributed tables inside of transactions, but we always use the Read Committed isolation level.

Given a schema:

```
CREATE TABLE first (key INT, value INT);
CREATE TABLE local_first (key INT, value INT);
SELECT master_create_distributed_table('first', 'key', 'hash');
SELECT master_create_worker_shards('first', 4, 2);
```

**What postgres does:**

```
-- session 1
brian=# BEGIN;
BEGIN
brian=# SET TRANSACTION ISOLATION LEVEL REPEATABLE READ ;
SET
brian=# SELECT * FROM local_first;
 key | value 
-----+-------
(0 rows)

-- now open session 2
brian=# INSERT INTO local_first VALUES (1, 1);
INSERT 0 1

-- now back to session 1
brian=# SELECT * FROM local_first;
 key | value 
-----+-------
(0 rows)

brian=# COMMIT;
COMMIT
brian=# SELECT * FROM local_first;
 key | value 
-----+-------
   1 |     1
(1 row)
```

No modifications made by transactions which finished after the first transaction started are visible to that first transaction.

**What Citus does**:

```
-- session 1
brian=# BEGIN;
BEGIN
brian=# SET TRANSACTION ISOLATION LEVEL REPEATABLE READ ;
SET
brian=# SELECT * FROM first ;
 key | value 
-----+-------
(0 rows)

-- now session 2
brian=# INSERT INTO first VALUES (1, 1);
INSERT 0 1

-- and session 1 again
brian=# SELECT * FROM first ;
 key | value 
-----+-------
   1 |     1
(1 row)
```

Citus shows data modifications which were made in other sessions, even though the transaction is still open.

**Proposed Fix**: I think we should disallow transactions on distributed tables while in either REPEATABLE READ or SERIALIZABLE

**Comment by anarazel** [comment]
We don't support repeatable read for reads either, and that's not new.

**Comment by lithp** [comment]
I don't think the problem here is supporting repeatable reads, but implicitly claiming that we do by running queries after Postgres has been explicitly asked to use that isolation level.

**Comment by marcocitus** [comment]
It's tricky. For instance, repeatable read would require us to reserve a snapshot on all the nodes, but we would rather only interact with a single node when dealing with a single shard transaction to maintain reasonable level of scalability. There are some more advanced techniques, but they require deeper changes to postgres and still have scalability issues.

**Comment by lorenzo** [comment]
What about transactions that only deal with a single node? Is that already supported?

**Comment by lorenzo** [comment]
@marcocitus sorry for the direct mention, but I’m currently evaluating city’s for my company and I cannot find any information in the docs about support for isolation levels when hitting a single node. Is this something you can help me clarify?

**Comment by marcocitus** [comment]
Currently, we explicitly set "read committed" (the default in Postgres) and don't support other isolation levels. When doing single-node transactions, the read committed isolation guarantees are the same as in PostgreSQL, since we simply delegate the full transaction to a worker node.

The main anomaly we have compared to PostgreSQL is that there is no distributed snapshot isolation for multi-shard queries, meaning a multi-shard query might see the effects of a concurrent multi-shard transaction that is in the process of committing on one node, but not on another. When there are no/few multi-shard transactions, this is mostly irrelevant.

**Comment by lorenzo** [comment]
It is relevant for me since I am expecting to send multiple concurrent transactions to the same shard and my application relies on an isolation level that is stricter than "read committed".

Thanks for your reply, @marcocitus

**Comment by slava-in** [comment]
we also hit similar problem when trying to pick new events from staging table while the ingest process was still writing to that table. Without "read committed" isolation, we had to complicate the process by additional update step that causes us performance loss.

**Comment by onderkalaci** [comment]
For `single-node` commands, this could work via 

```SQL
SET citus.propagate_set_commands TO 'local';
BEGIN ;
SET LOCAL transaction_read_only TO on;
SET TRANSACTION ISOLATION LEVEL REPEATABLE READ;
...
```

See https://github.com/citusdata/citus/pull/4945/files for details

**Comment by marcocitus** [comment]
Fixed by #6483
